vesc_hw_interface/src/vesc_servo_controller.cpp

```diff
@@ -74,60 +81,89 @@ void VescServoController::init(ros::NodeHandle nh, VescInterface* interface_ptr)
   {
     ROS_ERROR("[Servo Calibration] Invalid mode");
   }
+
+  // Smoothing differentiation when hall sensor resolution is insufficient
+  bool smooth_diff;
+  nh.param<bool>("servo/enable_smooth_diff", smooth_diff, true);
+  if (smooth_diff)
+  {
+    double smooth_diff_max_sampling_time;
+    int counter_td_vw_max_step;
+    nh.param<double>("servo/smooth_diff/max_sample_sec", smooth_diff_max_sampling_time, 1.0);
+    nh.param<int>("servo/smooth_diff/max_smooth_step", counter_td_vw_max_step, 10);
+    vesc_step_difference_.enableSmooth(control_rate_, smooth_diff_max_sampling_time, counter_td_vw_max_step);
+    ROS_INFO("[Motor Control] Smooth differentiation enabled, max_sample_sec: %f, max_smooth_step: %d",
+             smooth_diff_max_sampling_time, counter_td_vw_max_step);
+  }
   // Create timer callback for PID servo control
-  control_timer_ = nh.createTimer(ros::Duration(control_period_), &VescServoController::controlTimerCallback, this);
+  control_timer_ = nh.createTimer(ros::Duration(1.0 / control_rate_), &VescServoController::controlTimerCallback, this);
   return;
 }
 
-void VescServoController::control(const double position_reference, const double position_current)
+void VescServoController::control()
 {
   // executes calibration
   if (calibration_flag_)
   {
-    calibrate(position_current);
+    calibrate(sens_pose_);
     // initializes/resets control variables
-    time_previous_ = ros::Time::now();
-    position_sens_previous_ = position_current;
-    position_reference_previous_ = calibration_position_;
-    error_previous_ = 0.0;
+    sens_pose_previous_ = sens_pose_;
+    target_pose_previous_ = calibration_position_;
     return;
   }
 
-  const ros::Time time_current = ros::Time::now();
-  // calculates PD control
-  const double error_current = position_reference - position_current;
-  const double u_pd = Kp_ * error_current + Kd_ * (error_current - error_previous_) / control_period_;
+  // calculates PID control
+  double step_diff = vesc_step_difference_.getStepDifference(position_steps_, false);
+  double current_vel = step_diff * 2.0 * M_PI / (num_rotor_poles_ * num_hall_sensors_) * control_rate_ * gear_ratio_;
+  double target_vel = (target_pose_limited_ - target_pose_previous_) / control_rate_;
```

Comment:
Correct me if I'm wrong but shouldn't this be divided by `control_period_` and not `control_rate_`?

Reply:
I made a mistake in division and multiplication.
I will correct this.